File: solutions/zig/binary-search/1/binary_search.zig
// Take a look at the tests, you might have to change the function arguments

pub fn binarySearch(comptime T: type, target: T, items: []const T) ?usize {
    var left: usize = 0;
    var right = items.len;
    if (right == 0) {
        return null;
    }
    while (left + 1 < right) {
        const middle = (left + right) / 2;
        if (items[middle] <= target) {
            left = middle;
        } else {
            right = middle;
        }
    }
    return if (items[left] == target) left else null;
}

File: solutions/kotlin/binary-search/1/src/main/kotlin/BinarySearch.kt
object BinarySearch {
    fun search(list: List<Int>, item: Int): Int {
        var left = 0
        var right = list.size - 1
        while (left < right) {
            val middle = (left + right + 1) / 2
            if (list[middle] <= item) {
                left = middle
            } else {
                right = middle - 1
            }
        }
        require(list[left] == item)
        return left
    }
}
